android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/NearbyBeaconsFragment.java

```diff
@@ -337,16 +345,57 @@ private void emptyGroupIdQueue() {
 
     for (String groupId : mGroupIdQueue) {
       Log.d(TAG, "groupid " + groupId);
-      pwPairs.add(Utils.getTopRankedPwPairByGroupId(mPwCollection, groupId));
+      pwPairs.add(Utils.getTopRankedPwPairByGroupId(mPwCollection, groupId, mFavorites));
     }
-    Collections.sort(pwPairs, Utils.newDistanceComparator());
+    Collections.sort(pwPairs, Utils.newDistanceComparator(mFavorites));
     for (PwPair pwPair : pwPairs) {
       mNearbyDeviceAdapter.addItem(pwPair);
     }
     mGroupIdQueue.clear();
     mNearbyDeviceAdapter.notifyDataSetChanged();
   }
 
+  public boolean getFavorite(String siteUrl) {
+    return mFavorites.contains(siteUrl);
+  }
+
+  public void flipFavorite(String siteUrl) {
+    if (getFavorite(siteUrl)) {
+      mFavorites.remove(siteUrl);
+      return;
+    }
+    mFavorites.add(siteUrl);
+  }
+
+  private void saveFavorites() {
+    // Write the PW Collection
+    String preferencesKey = getString(R.string.discovery_service_prefs_key);
+    SharedPreferences prefs = getActivity().getSharedPreferences(
+        preferencesKey, Context.MODE_PRIVATE);
+    SharedPreferences.Editor editor = prefs.edit();
+    try {
+      editor.putString("fav", (new JSONArray(mFavorites.toArray())).toString());
+    } catch (JSONException e) {
+      Log.e(TAG, "Could not restore Physical Web collection cache", e);
+    }
+    editor.apply();
+  }
+
+  private void getFavorites() {
```

Comment:
"getFavorites" suggests that this function returns the list of favorites, could you rename this "restoreFavorites"?